## Intoduction to the Data

#### The Dataset

Exhaustive one year non-confidential data in the year 2015 to 2016 of firms is collected from the Auditor Office of India to be used in building a predictor for classifying suspicious firms.


The columns (features) in the dataset consists of many risk factors that are examined from various areas like past records of audit office, audit-paras, environmental conditions reports, firm reputation summary, on-going issues report, profit-value records, loss-value records, follow-up reports etc. 

After in-depth interview with the auditors, important risk factors are evaluated and their probability of existence is calculated from the present and past records.



The goal of the research is to help the auditors by building a classification model that can predict the fraudulent firm on the basis the present and historical risk factors.

In this project, I'll:

- Prepare the data for machine learning
- Train a model that can predict the fraudulent firm Using a Logistic Regression model
- Measure the accuracy of the model

In [1]:
import pandas as pd
audit_risk = pd.read_csv("audit_risk.csv", encoding="Utf-8")
audit_risk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 776 entries, 0 to 775
Data columns (total 27 columns):
Sector_score      776 non-null float64
LOCATION_ID       776 non-null object
PARA_A            776 non-null float64
Score_A           776 non-null float64
Risk_A            776 non-null float64
PARA_B            776 non-null float64
Score_B           776 non-null float64
Risk_B            776 non-null float64
TOTAL             776 non-null float64
numbers           776 non-null float64
Score_B.1         776 non-null float64
Risk_C            776 non-null float64
Money_Value       775 non-null float64
Score_MV          776 non-null float64
Risk_D            776 non-null float64
District_Loss     776 non-null int64
PROB              776 non-null float64
RiSk_E            776 non-null float64
History           776 non-null int64
Prob              776 non-null float64
Risk_F            776 non-null float64
Score             776 non-null float64
Inherent_Risk     776 non-null float64
CON

In [2]:
audit_risk.head(50)

Sector_score LOCATION_ID  PARA_A  Score_A  Risk_A  PARA_B  Score_B  \
0           3.89          23    4.18      0.6   2.508    2.50      0.2   
1           3.89           6    0.00      0.2   0.000    4.83      0.2   
2           3.89           6    0.51      0.2   0.102    0.23      0.2   
3           3.89           6    0.00      0.2   0.000   10.80      0.6   
4           3.89           6    0.00      0.2   0.000    0.08      0.2   
5           3.89           6    0.00      0.2   0.000    0.83      0.2   
6           3.89           7    1.10      0.4   0.440    7.41      0.4   
7           3.89           8    8.50      0.6   5.100   12.03      0.6   
8           3.89           8    8.40      0.6   5.040   11.05      0.6   
9           3.89           8    3.98      0.6   2.388    0.99      0.2   
10          3.89           8    5.43      0.6   3.258   10.77      0.6   
11          3.89           8   15.38      0.6   9.228   40.14      0.6   
12          3.89           8    5.47      0.6   3.282    7.63      0.4   
13          3.89           8    1.09      0.4   0.436    0.35      0.2   
14          3.89           8    0.00      0.2   0.000    0.84      0.2   
15          3.89          13    1.95      0.4   0.780    9.01      0.4   
16          3.89          37    8.54      0.6   5.124   31.63      0.6   
17          3.89          37    4.18      0.6   2.508    4.83      0.2   
18          3.89          37    1.81      0.4   0.724    1.03      0.2   
19          3.89          37    4.86      0.6   2.916   46.78      0.6   
20          3.89          24    6.26      0.6   3.756   14.10      0.6   
21          3.89           3    0.02      0.2   0.004    5.94      0.4   
22          3.89           3    5.31      0.6   3.186   22.79      0.6   
23          3.89           3    0.94      0.2   0.188    0.01      0.2   
24          3.89           4    5.78      0.6   3.468   57.92      0.6   
25          3.89           4    7.42      0.6   4.452    2.24      0.2   
26          3.89           4    0.00      0.2   0.000    1.10      0.2   
27          3.89          14    6.85      0.6   4.110   31.76      0.6   
28          3.89          14    0.00      0.2   0.000    1.03      0.2   
29          3.89          37    0.00      0.2   0.000    0.75      0.2   
30          3.89          37    2.40      0.6   1.440   16.63      0.6   
31          3.89           5    0.00      0.2   0.000    0.05      0.2   
32          3.89           5    0.00      0.2   0.000    1.76      0.2   
33          3.89           5    0.00      0.2   0.000    2.97      0.2   
34          3.89           5    0.00      0.2   0.000    0.43      0.2   
35          3.89           5    0.00      0.2   0.000    0.94      0.2   
36          3.89          20    9.01      0.6   5.406   19.82      0.6   
37          3.89          19    0.00      0.2   0.000    0.05      0.2   
38          3.89          19   11.95      0.6   7.170   30.90      0.6   
39          3.89          19    7.97      0.6   4.782   17.18      0.6   
40          3.89          19    0.00      0.2   0.000    3.71      0.2   
41          3.89          19    0.62      0.2   0.124    2.02      0.2   
42          3.89          19    0.00      0.2   0.000    2.27      0.2   
43          3.89          19    0.00      0.2   0.000    0.00      0.2   
44          3.89          19    1.32      0.4   0.528    0.00      0.2   
45          3.89          19    0.00      0.2   0.000    1.03      0.2   
46          3.89          21    0.00      0.2   0.000   15.38      0.6   
47          3.89          21    0.00      0.2   0.000    1.02      0.2   
48          3.89          21    0.00      0.2   0.000   20.71      0.6   
49          3.89          22    2.95      0.6   1.770    6.56      0.4   

    Risk_B  TOTAL  numbers  ...  RiSk_E  History  Prob  Risk_F  Score  \
0    0.500   6.68      5.0  ...     0.4        0   0.2     0.0    2.4   
1    0.966   4.83      5.0  ...     0.4        0   0.2     0.0    2.0   
2    0.046

## Feature Selection

Selecting the Risk Columnn as the target column

In [4]:
for Col in audit_risk.columns:
    print(audit_risk[Col].value_counts(normalize=True)*100)

55.57    25.773196
3.89     14.690722
1.85     12.242268
2.72     10.567010
3.41      9.793814
2.37      9.536082
1.99      6.056701
21.61     5.283505
59.85     4.768041
2.34      0.644330
15.56     0.386598
2.36      0.128866
17.68     0.128866
Name: Sector_score, dtype: float64
8          9.793814
19         8.762887
9          6.829897
16         6.701031
12         6.056701
5          5.670103
2          5.283505
4          4.768041
15         4.510309
13         4.510309
6          4.252577
32         3.737113
11         3.350515
22         3.092784
29         2.706186
14         2.577320
18         2.061856
31         1.546392
1          1.417526
37         1.288660
39         1.159794
27         1.030928
21         1.030928
28         1.030928
43         0.902062
25         0.773196
20         0.644330
36         0.515464
30         0.515464
7          0.515464
38         0.515464
3          0.386598
40         0.386598
35         0.257732
41         0.128866
NUH        0.12886

I'll drop the Detection_Risk column, 'cause it contains only one unique value - which does not give us any information

In [5]:
audit_risk = audit_risk.drop(["Detection_Risk"], axis=1)

## Finding Missing Values

In [6]:
audit_risk.isnull().sum()

Sector_score     0
LOCATION_ID      0
PARA_A           0
Score_A          0
Risk_A           0
PARA_B           0
Score_B          0
Risk_B           0
TOTAL            0
numbers          0
Score_B.1        0
Risk_C           0
Money_Value      1
Score_MV         0
Risk_D           0
District_Loss    0
PROB             0
RiSk_E           0
History          0
Prob             0
Risk_F           0
Score            0
Inherent_Risk    0
CONTROL_RISK     0
Audit_Risk       0
Risk             0
dtype: int64

In [7]:
audit_risk = audit_risk.dropna(axis=0)

In [8]:
audit_risk.isnull().sum()

Sector_score     0
LOCATION_ID      0
PARA_A           0
Score_A          0
Risk_A           0
PARA_B           0
Score_B          0
Risk_B           0
TOTAL            0
numbers          0
Score_B.1        0
Risk_C           0
Money_Value      0
Score_MV         0
Risk_D           0
District_Loss    0
PROB             0
RiSk_E           0
History          0
Prob             0
Risk_F           0
Score            0
Inherent_Risk    0
CONTROL_RISK     0
Audit_Risk       0
Risk             0
dtype: int64

## Data  and Feature Engineering
 
Since PARA_A and PARA_B constitutes of audit_paras done more than once

Same for:
-  Score and Score B 
- Risk_A - Risk_F

To get a more nuanced data, I'll add up the respective columns for a particular feature

In [10]:
audit_risk.columns

Index(['Sector_score', 'LOCATION_ID', 'PARA_A', 'Score_A', 'Risk_A', 'PARA_B',
       'Score_B', 'Risk_B', 'TOTAL', 'numbers', 'Score_B.1', 'Risk_C',
       'Money_Value', 'Score_MV', 'Risk_D', 'District_Loss', 'PROB', 'RiSk_E',
       'History', 'Prob', 'Risk_F', 'Score', 'Inherent_Risk', 'CONTROL_RISK',
       'Audit_Risk', 'Risk'],
      dtype='object')

In [11]:
audit_risk["Para_Total"] = audit_risk["PARA_A"] +  audit_risk["PARA_B"]
audit_risk["Score_Total"] =  audit_risk["Score_A"] + audit_risk["Score_B"] 
audit_risk["Risk_Total"] =  audit_risk["Risk_A"] + audit_risk["Risk_B"] + audit_risk["Risk_C"] + audit_risk["Risk_D"] + audit_risk["RiSk_E"] + audit_risk["Risk_F"] 

In [12]:
audit_risk = audit_risk.drop(['PARA_A', 'Score_A', 'Risk_A', 'PARA_B',
       'Score_B', 'Risk_B','Risk_D','Risk_C','Risk_F','RiSk_E'], axis=1)

In [13]:
audit_risk

Sector_score LOCATION_ID    TOTAL  numbers  Score_B.1  Money_Value  \
0            3.89          23   6.6800      5.0        0.2        3.380   
1            3.89           6   4.8300      5.0        0.2        0.940   
2            3.89           6   0.7400      5.0        0.2        0.000   
3            3.89           6  10.8000      6.0        0.6       11.750   
4            3.89           6   0.0800      5.0        0.2        0.000   
5            3.89           6   0.8300      5.0        0.2        2.950   
6            3.89           7   8.5100      5.0        0.2       44.950   
7            3.89           8  20.5300      5.5        0.4        7.790   
8            3.89           8  19.4500      5.5        0.4        7.340   
9            3.89           8   4.9700      5.0        0.2        1.930   
10           3.89           8  16.2000      5.0        0.2        4.420   
11           3.89           8  55.5200      5.0        0.2        0.960   
12           3.89           8  13.1000      5.0        0.2       10.430   
13           3.89           8   1.4400      5.0        0.2        0.000   
14           3.89           8   0.8400      5.0        0.2        0.007   
15           3.89          13  10.9600      5.0        0.2        9.000   
16           3.89          37  40.1700      5.0        0.2       41.280   
17           3.89          37   9.0100      5.5        0.4       14.030   
18           3.89          37   2.8400      5.0        0.2        0.000   
19           3.89          37  51.6400      5.5        0.4       63.180   
20           3.89          24  20.3600      5.0        0.2       34.240   
21           3.89           3   5.9600      5.0        0.2        0.010   
22           3.89           3  28.1000      5.0        0.2      205.190   
23           3.89           3   0.9500      5.0        0.2        0.100   
24           3.89           4  63.7000      5.0        0.2       11.160   
25           3.89           4   9.6600      5.0        0.2        1.250   
26           3.89           4   1.1000      5.0        0.2        0.007   
27           3.89          14  38.6100      5.0        0.2        1.460   
28           3.89          14   1.0300      5.0        0.2        0.000   
29           3.89          37   0.7500      5.0        0.2        6.780   
..            ...         ...      ...      ...        ...          ...   
746         55.57          13   0.2517      5.0        0.2        0.000   
747         55.57          13   0.3115      5.0        0.2        0.000   
748         55.57          13   0.0000      5.0        0.2        0.000   
749         55.57          13   0.8400      5.0        0.2        0.000   
750         55.57          13   1.0900      5.0        0.2        0.000   
751         55.57          13   1.2900      5.0        0.2        0.000   
752         55.57          13   0.8800      5.0        0.2        0.000   
753         55.57          21   0.0900      5.0        0.2        0.000   
754         55.57          18   1.3000      5.0        0.2        0.000   
755         55.57          21   1.0700      5.0        0.2        0.000   
756         55.57          25   0.0000      5.0        0.2        0.000   
757         55.57          32   3.4700      5.0        0.2        0.180   
758         55.57          22   1.0400      5.0        0.2        0.000   
759         55.57          14   1.4900      5.0        0.2        0.000   
760         55.57          12   2.0100      5.0        0.2        0.000   
761         55.57          12   0.0000      5.0        0.2        0.000   
762         55.57          14   0.5900      5.0        0.2        0.000   
763         55.57          36   0.0200      5.0        0.2        0.000   
764         55.57          14   5.9600      5.0        0.2        0.000   
765         55.57          22   3.3000      5.0        0.2        0.000   
766         55.57           8   1.3700      5.0        0.2        0.000   
767         55.57          18 

#### Creating Dummies

The LOCATION_ID column is a categorical column (object)

In [15]:
cat_columns = ['LOCATION_ID']
dummy_df = pd.get_dummies(audit_risk[cat_columns])
audit_risk = pd.concat([audit_risk, dummy_df], axis=1)
audit_risk = audit_risk.drop(cat_columns, axis=1)

In [14]:
audit_risk.corr()["Risk"]

Sector_score    -0.393322
TOTAL            0.291893
numbers          0.308017
Score_B.1        0.353664
Money_Value      0.256992
Score_MV         0.688207
District_Loss    0.403591
PROB             0.176818
History          0.239354
Prob             0.298520
Score            0.785813
Inherent_Risk    0.356881
CONTROL_RISK     0.416285
Audit_Risk       0.217023
Risk             1.000000
Para_Total       0.292676
Score_Total      0.707529
Risk_Total       0.356773
Name: Risk, dtype: float64

## Training a Model with a Logistic Regression Model

From the code block, the feature columns are strongly correlated with the target column, hence a Logistic Regression Model is the right call.

I'll be using accuracy as my error metric

In [17]:
audit_risk_features = audit_risk.drop(["Risk"], axis=1)

In [18]:
from warnings import simplefilter
# ignore all future warnings
simplefilter(action='ignore', category=FutureWarning)

### Selecting The Most Relevant columns using RFECV

In [20]:
from sklearn.feature_selection import RFECV
from sklearn.linear_model import LogisticRegression


cols = audit_risk_features.columns
all_X = audit_risk_features[cols]
all_y = audit_risk["Risk"]
lr = LogisticRegression()
selector = RFECV(lr,cv=10)
selector.fit(all_X,all_y)

optimized_columns = all_X.columns[selector.support_]

In [21]:
from sklearn.model_selection import cross_val_score
import numpy as np
X = audit_risk_features[optimized_columns]
Y = audit_risk["Risk"]

lr = LogisticRegression(class_weight="balanced")
scores = cross_val_score(lr, X, Y, cv=10)
accuracy = np.mean(scores)
print(accuracy)

0.9948218448218448


The Classification Model for Predicting Suspicious Firms had an impressive performance of 99.48%